File: scripts/create-compact-embeddings.mjs
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Path to the full embeddings file
const fullEmbeddingsPath = path.join(__dirname, '../src/data/glove_embeddings_3d_full.json');
const outputPath = path.join(__dirname, '../src/data/embeddings_with_words.json');

// Common English words that should be prioritized
const PRIORITY_WORDS = new Set([
    // Core function words
    'the', 'be', 'to', 'of', 'and', 'a', 'in', 'that', 'have', 'i', 'it', 'for', 'not', 'on', 'with', 'he', 'as', 'you', 'do', 'at',
    'this', 'but', 'his', 'by', 'from', 'they', 'we', 'say', 'her', 'she', 'or', 'an', 'will', 'my', 'one', 'all', 'would', 'there', 'their',
    'what', 'up', 'out', 'if', 'about', 'who', 'get', 'which', 'go', 'me', 'when', 'make', 'can', 'like', 'time', 'no', 'just', 'him', 'know', 'take',
    
    // Common nouns
    'man', 'woman', 'person', 'people', 'child', 'children', 'family', 'friend', 'house', 'home', 'car', 'book', 'water', 'food', 'money', 'work', 'job',
    'school', 'student', 'teacher', 'doctor', 'city', 'country', 'world', 'life', 'death', 'love', 'peace', 'war', 'music', 'art', 'science', 'nature',
    'animal', 'plant', 'tree', 'flower', 'mountain', 'river', 'ocean', 'beach', 'forest', 'desert', 'sky', 'sun', 'moon', 'star', 'earth', 'fire', 'air',
    
    // Common verbs
    'see', 'hear', 'feel', 'think', 'know', 'understand', 'remember', 'forget', 'learn', 'teach', 'read', 'write', 'speak', 'listen', 'walk', 'run',
    'eat', 'drink', 'sleep', 'wake', 'open', 'close', 'start', 'stop', 'come', 'go', 'give', 'take', 'put', 'get', 'make', 'do', 'work', 'play',
    'help', 'ask', 'answer', 'tell', 'show', 'look', 'watch', 'find', 'lose', 'win', 'try', 'want', 'need', 'like', 'love', 'hate', 'hope', 'fear',
    
    // Common adjectives
    'good', 'bad', 'big', 'small', 'large', 'little', 'old', 'new', 'young', 'hot', 'cold', 'warm', 'cool', 'fast', 'slow', 'high', 'low', 'long', 'short',
    'happy', 'sad', 'angry', 'excited', 'tired', 'hungry', 'thirsty', 'sick', 'healthy', 'strong', 'weak', 'smart', 'stupid', 'beautiful', 'ugly',
    'easy', 'hard', 'simple', 'difficult', 'important', 'interesting', 'boring', 'funny', 'serious', 'safe', 'dangerous', 'free', 'expensive', 'cheap',
    
    // Colors
    'red', 'blue', 'green', 'yellow', 'orange', 'purple', 'pink', 'brown', 'black', 'white', 'gray', 'grey',
    
    // Numbers
    'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten', 'hundred', 'thousand', 'million',
    
    // Animals
    'cat', 'dog', 'bird', 'fish', 'horse', 'cow', 'pig', 'sheep', 'chicken', 'mouse', 'rat', 'lion', 'tiger', 'bear', 'elephant', 'monkey', 'rabbit',
    
    // Food
    'apple', 'banana', 'orange', 'bread', 'milk', 'cheese', 'meat', 'chicken', 'fish', 'rice', 'pasta', 'pizza', 'cake', 'chocolate', 'coffee', 'tea',
    
    // Body parts
    'head', 'face', 'eye', 'eyes', 'nose', 'mouth', 'ear', 'ears', 'hair', 'hand', 'hands', 'finger', 'foot', 'feet', 'leg', 'legs', 'arm', 'arms',
    
    // Technology
    'computer', 'phone', 'internet', 'email', 'game', 'music', 'video', 'camera', 'television', 'radio', 'machine',
    
    // Places
    'store', 'shop', 'restaurant', 'hotel', 'hospital', 'airport', 'station', 'park', 'street', 'road', 'bridge', 'building',
    
    // Time
    'today', 'yesterday', 'tomorrow', 'morning', 'afternoon', 'evening', 'night', 'day', 'week', 'month', 'year', 'hour', 'minute', 'second',
    'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday', 'sunday',
    'january', 'february', 'march', 'april', 'may', 'june', 'july', 'august', 'september', 'october', 'november', 'december'
]);

function isValidWord(word) {
    // Filter out words that are likely not useful for the game
    if (word.length < 2 || word.length > 15) return false;
    if (!word.match(/^[a-z]+$/)) return false; // Only lowercase letters
    if (word.includes('_') || word.includes('-')) return false;
    
    // Filter out very technical or obscure words
    const technicalPrefixes = ['bio', 'geo', 'neo', 'proto', 'pseudo', 'semi', 'anti', 'pre', 'post', 'non'];
    const technicalSuffixes = ['ology', 'ism', 'ist', 'tion', 'sion', 'ness', 'ment', 'able', 'ible'];
    
    // Allow some technical words but not overly complex ones
    if (word.length > 10) {
        const hasTechnicalAffix = technicalPrefixes.some(prefix => word.startsWith(prefix)) ||
                                  technicalSuffixes.some(suffix => word.endsWith(suffix));
        if (hasTechnicalAffix) return false;
    }
    
    return true;
}

function scoreWord(word) {
    let score = 0;
    
    // Priority words get highest score
    if (PRIORITY_WORDS.has(word)) {
        score += 1000;
    }
    
    // Prefer shorter, more common words
    if (word.length <= 5) score += 100;
    else if (word.length <= 7) score += 50;
    else if (word.length <= 10) score += 10;
    
    // Boost common word patterns
    const commonEndings = ['ing', 'ed', 'er', 'ly', 'tion', 'ness'];
    if (commonEndings.some(ending => word.endsWith(ending))) {
        score += 20;
    }
    
    // Vowel distribution (words with good vowel distribution are often more "real")
    const vowels = (word.match(/[aeiou]/g) || []).length;
    const vowelRatio = vowels / word.length;
    if (vowelRatio >= 0.2 && vowelRatio <= 0.6) {
        score += 30;
    }
    
    return score;
}

async function createCompactEmbeddings() {
    console.log('📥 Loading full embeddings file...');
    
    if (!fs.existsSync(fullEmbeddingsPath)) {
        console.error(`❌ Full embeddings file not found at ${fullEmbeddingsPath}`);
        console.log('Please make sure you have the full GloVe embeddings file.');
        return;
    }
    
    const fullEmbeddings = JSON.parse(fs.readFileSync(fullEmbeddingsPath, 'utf8'));
    console.log(`📊 Loaded ${fullEmbeddings.length} word embeddings`);
    
    console.log('🔍 Filtering and scoring words...');
    
    // Filter and score words
    const validEmbeddings = fullEmbeddings
        .filter(embedding => isValidWord(embedding.word))
        .map(embedding => ({
            ...embedding,
            score: scoreWord(embedding.word)
        }))
        .sort((a, b) => b.score - a.score) // Sort by score descending
        .slice(0, 20000) // Take top 20,000
        .map(embedding => {
            // Remove the score field from final output
            const { score, ...rest } = embedding;
            return rest;
        });
    
    console.log('💾 Saving compact embeddings...');
    fs.writeFileSync(outputPath, JSON.stringify(validEmbeddings, null, 0));
    
    // Check file sizes
    const originalStats = fs.statSync(fullEmbeddingsPath);
    const newStats = fs.statSync(outputPath);
    const originalSize = originalStats.size / (1024 * 1024); // MB
    const newSize = newStats.size / (1024 * 1024); // MB
    
    console.log('✅ Created compact embeddings file!');
    console.log(`📊 Original size: ${originalSize.toFixed(1)} MB`);
    console.log(`📊 New size: ${newSize.toFixed(1)} MB`);
    console.log(`📊 Reduction: ${((originalSize - newSize) / originalSize * 100).toFixed(1)}%`);
    console.log(`📊 Words included: ${validEmbeddings.length}`);
    
    // Show some sample words
    const sampleWords = validEmbeddings.slice(0, 20).map(emb => emb.word);
    console.log(`📝 Sample words: ${sampleWords.join(', ')}`);
    
    // Show priority words included
    const priorityIncluded = validEmbeddings
        .filter(emb => PRIORITY_WORDS.has(emb.word))
        .length;
    console.log(`🎯 Priority words included: ${priorityIncluded}/${PRIORITY_WORDS.size}`);
}

createCompactEmbeddings().catch(console.error);

File: scripts/create-compact-embeddings.py
#!/usr/bin/env python3
import json
import os
from collections import Counter

# Path to the full embeddings file
full_embeddings_path = '../src/data/glove_embeddings_3d_full.json'
output_path = '../src/data/embeddings_with_words.json'

# List of most common English words (you can expand this list)
# These are based on frequency in common usage
COMMON_WORDS = [
    # Top function words
    'the', 'be', 'to', 'of', 'and', 'a', 'in', 'that', 'have', 'i', 'it', 'for', 'not', 'on', 'with', 'he', 'as', 'you', 'do', 'at',
    'this', 'but', 'his', 'by', 'from', 'they', 'we', 'say', 'her', 'she', 'or', 'an', 'will', 'my', 'one', 'all', 'would', 'there', 'their',
    
    # Common nouns
    'time', 'person', 'year', 'way', 'day', 'thing', 'man', 'world', 'life', 'hand', 'part', 'child', 'eye', 'woman', 'place', 'work', 'week', 'case', 'point', 'government',
    'company', 'number', 'group', 'problem', 'fact', 'money', 'business', 'service', 'book', 'water', 'food', 'house', 'car', 'school', 'family', 'state', 'country', 'area',
    
    # Common verbs
    'get', 'go', 'make', 'know', 'take', 'see', 'come', 'think', 'look', 'want', 'give', 'use', 'find', 'tell', 'ask', 'work', 'seem', 'feel', 'try', 'leave',
    'call', 'good', 'new', 'first', 'last', 'long', 'great', 'little', 'own', 'other', 'old', 'right', 'big', 'high', 'different', 'small', 'large', 'next', 'early', 'young',
    
    # Common adjectives
    'important', 'few', 'public', 'bad', 'same', 'able', 'local', 'sure', 'possible', 'late', 'hard', 'far', 'major', 'better', 'economic', 'strong', 'necessary', 'clear', 'real', 'trade',
    
    # Colors
    'red', 'blue', 'green', 'yellow', 'black', 'white', 'brown', 'orange', 'purple', 'pink', 'gray', 'grey',
    
    # Animals
    'cat', 'dog', 'bird', 'fish', 'horse', 'cow', 'pig', 'chicken', 'sheep', 'mouse', 'lion', 'tiger', 'bear', 'elephant', 'monkey', 'rabbit', 'deer', 'wolf',
    
    # Food
    'apple', 'banana', 'orange', 'bread', 'milk', 'cheese', 'meat', 'chicken', 'fish', 'rice', 'pasta', 'pizza', 'cake', 'chocolate', 'coffee', 'tea', 'beer', 'wine',
    
    # Body parts
    'head', 'face', 'eye', 'nose', 'mouth', 'ear', 'hair', 'neck', 'shoulder', 'arm', 'hand', 'finger', 'leg', 'foot', 'back', 'chest', 'heart', 'brain',
    
    # Nature
    'sun', 'moon', 'star', 'sky', 'cloud', 'rain', 'snow', 'wind', 'tree', 'flower', 'grass', 'mountain', 'river', 'lake', 'ocean', 'beach', 'forest', 'desert',
    
    # Technology
    'computer', 'phone', 'internet', 'website', 'email', 'software', 'app', 'game', 'music', 'video', 'camera', 'television', 'radio', 'machine', 'robot',
    
    # Emotions
    'happy', 'sad', 'angry', 'excited', 'worried', 'surprised', 'scared', 'proud', 'lonely', 'grateful', 'confused', 'disappointed', 'nervous', 'calm', 'peaceful'
]

def load_word_frequency_list():
    """
    Load a more comprehensive list of common words.
    In a real implementation, you'd load this from a frequency corpus like Google's n-grams.
    For now, we'll use a basic approach.
    """
    # This is a simplified approach - in practice you'd use a real frequency list
    return set(COMMON_WORDS)

def filter_embeddings_by_frequency(embeddings_data, max_words=20000):
    """
    Filter embeddings to keep only the most common/useful words.
    """
    print(f"Starting with {len(embeddings_data)} word embeddings")
    
    # Get our priority words
    priority_words = load_word_frequency_list()
    
    # Separate into priority and non-priority
    priority_embeddings = []
    other_embeddings = []
    
    for embedding in embeddings_data:
        word = embedding['word'].lower()
        if word in priority_words:
            priority_embeddings.append(embedding)
        else:
            # Filter out very uncommon words or problematic ones
            if (len(word) >= 2 and 
                len(word) <= 15 and 
                word.isalpha() and 
                not any(char in word for char in ['_', '-', '.', "'"])):
                other_embeddings.append(embedding)
    
    print(f"Found {len(priority_embeddings)} priority words")
    print(f"Found {len(other_embeddings)} other valid words")
    
    # Take all priority words plus fill up to max_words with others
    remaining_slots = max_words - len(priority_embeddings)
    
    if remaining_slots > 0:
        # Sort others by word length and alphabetically for consistency
        other_embeddings.sort(key=lambda x: (len(x['word']), x['word']))
        selected_others = other_embeddings[:remaining_slots]
    else:
        selected_others = []
    
    final_embeddings = priority_embeddings + selected_others
    
    print(f"Final selection: {len(final_embeddings)} words")
    return final_embeddings

def main():
    # Check if full embeddings file exists
    if not os.path.exists(full_embeddings_path):
        print(f"❌ Full embeddings file not found at {full_embeddings_path}")
        print("Please make sure you have the full GloVe embeddings file.")
        return
    
    print("📥 Loading full embeddings file...")
    with open(full_embeddings_path, 'r') as f:
        full_embeddings = json.load(f)
    
    print("🔍 Filtering to top 20,000 most useful words...")
    compact_embeddings = filter_embeddings_by_frequency(full_embeddings, max_words=20000)
    
    print("💾 Saving compact embeddings...")
    with open(output_path, 'w') as f:
        json.dump(compact_embeddings, f, separators=(',', ':'))
    
    # Check file sizes
    original_size = os.path.getsize(full_embeddings_path) / (1024 * 1024)  # MB
    new_size = os.path.getsize(output_path) / (1024 * 1024)  # MB
    
    print(f"✅ Created compact embeddings file!")
    print(f"📊 Original size: {original_size:.1f} MB")
    print(f"📊 New size: {new_size:.1f} MB")
    print(f"📊 Reduction: {((original_size - new_size) / original_size * 100):.1f}%")
    print(f"📊 Words included: {len(compact_embeddings)}")
    
    # Show some sample words
    sample_words = [emb['word'] for emb in compact_embeddings[:20]]
    print(f"📝 Sample words: {', '.join(sample_words)}")

if __name__ == "__main__":
    main()